ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,643 @@
+pub const NULL_AUTHOR: Address = H160([0x00; 20]);
+pub const NULL_NONCE: [u8; 8] = NONCE_DROP_VOTE;
+pub const NULL_MIXHASH: [u8; 32] = [0x00; 32];
+pub const NULL_UNCLES_HASH: H256 = KECCAK_EMPTY_LIST_RLP;
+pub const SIGNING_DELAY_NOTURN_MS: u64 = 500;
+
+#[derive(PartialEq, Clone, Debug, Copy)]
+pub enum VoteType {
+	Add,
+	Remove
+}
+
+// Caches
+/// How many CliqueBlockState to cache in the memory.
+pub const STATE_CACHE_NUM: usize = 128;
+
+/// Clique Engine implementation
+pub struct Clique {
+	epoch_length: u64,
+	period: u64,
+	machine: EthereumMachine,
+	client: RwLock<Option<Weak<EngineClient>>>,
+	block_state_by_hash: RwLock<LruCache<H256, CliqueBlockState>>,
+	proposals: RwLock<HashMap<Address, VoteType>>,
+	signer: RwLock<Option<Box<EngineSigner>>>,
+	step_service: IoService<Duration>,
+}
+
+impl Clique {
+	/// initialize Clique engine from empty state.
+	pub fn new(our_params: CliqueParams, machine: EthereumMachine) -> Result<Arc<Self>, Error> {
+		let engine = Arc::new(
+			Clique {
+				epoch_length: our_params.epoch,
+				period: our_params.period,
+				client: RwLock::new(Option::default()),
+				block_state_by_hash: RwLock::new(LruCache::new(STATE_CACHE_NUM)),
+				proposals: RwLock::new(Default::default()),
+				signer: RwLock::new(Default::default()),
+				machine,
+				step_service: IoService::<Duration>::start()?,
+			});
+
+		if engine.period > 0 {
+			let handler = StepService::new(Arc::downgrade(&engine) as Weak<Engine<_>>,
+			                               Duration::from_secs(our_params.period) / 2 + Duration::from_millis(10));
+			engine.step_service.register_handler(Arc::new(handler)).expect("set_signer: Error registering timeout handler.");
+		}
+
+		return Ok(engine);
+	}
+
+	fn sign_header(&self, header: &Header) -> Result<(Signature, H256), Error> {
+		match self.signer.read().as_ref() {
+			None => {
+				return Err(Box::new("sign_header: No signer available.").into());
+			}
+			Some(signer) => {
+				let digest = header.hash();
+				match signer.sign(digest) {
+					Ok(sig) => {
+						return Ok((sig, digest));
+					}
+					Err(e) => {
+						return Err(Box::new(format!("sign_header: failed to sign header, error: {}", e)).into());
+					}
+				}
+			}
+		}
+	}
+
+	/// Construct an new state from given checkpoint header.
+	fn new_checkpoint_state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		assert_eq!(header.number() % self.epoch_length, 0);
+
+		let mut state = CliqueBlockState::new(
+			match header.number() {
+				0 => NULL_AUTHOR,
+				_ => recover_creator(header)?,
+			},
+			extract_signers(header)?);
+
+		state.calc_next_timestamp(header, self.period);
+
+		Ok(state)
+	}
+
+	fn state_no_backfill(&self, hash: &H256) -> Option<CliqueBlockState> {
+		return self.block_state_by_hash.write().get_mut(hash).cloned()
+	}
+
+	/// get CliqueBlockState for given header, backfill from last checkpoint if needed.
+	fn state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
```

Comment:
I think this function can be broken down into smaller pieces. Also, there are a lot of calls to `.unwrap()` in there. They need to be removed and handled appropriately or justified through an `.expect()`.

Reply:
There are alot, but it is mostly due to how VecDec api works. All unrwap() calls is basically just after an push, which grauntee to not fail.  I'm not too sure how this could be changed

Reply:
See my other comment.

Reply:
fixed

Reply:
Do you think it is possible to split up this method even more?

It is quite annoying to clone/allocate `CliqueBlockState` even when just the cache is read.
It may be possible to leave this as it is to see how it performs and create an issue for this.